### PR TITLE
[next] Memoize lstating files

### LINKED ISSUE #14429: Vercel's Next.js builder should memoize lstats of files
To speed up the Next.js builder, we should memoize the lstat of files when bundling lambdas.

https://github.com/vercel/vercel/blob/83f77223aa34d3833981d0401158faab06754ac3/packages/now-next/src/index.ts#L925

### PR DESCRIPTION
**Pull request by ijjk**
Closes: https://github.com/vercel/next.js/issues/14429

### CODE REVIEW

**Review by Timer — DISMISSED** [review]

File: packages/now-next/src/index.ts [hunk 1]
```diff
@@ -913,25 +913,36 @@ export const build = async ({
...
 
         files[file] = new FileFsRef({
           fsPath: path.join(workPath, file),
           mode,
         });
       };
 
-      fileList.forEach(collectTracedFiles(nonApiReasons, tracedFiles));
-      apiFileList.forEach(collectTracedFiles(apiReasons, apiTracedFiles));
+      await Promise.all(
```
Comment (timneutkens) [inline packages/now-next/src/index.ts:947]: Does this need a semaphore in case there's loads of files?

File: packages/now-next/src/index.ts [hunk 2]
```diff
@@ -913,25 +913,36 @@ export const build = async ({
...
           // Initial files are manually added to the lambda later
           return;
         }
-        const { mode } = lstatSync(path.join(workPath, file));
+        const filePath = path.join(workPath, file);
+
+        if (!lstatResults[filePath]) {
+          lstatResults[filePath] = lstat(filePath);
+        }
+        const { mode } = await lstatResults[filePath];
```
Comment (Timer) [inline packages/now-next/src/index.ts:932]: Sema should be here, not in `Promise.all`.

File: packages/now-next/src/index.ts [hunk 3]
```diff
@@ -913,25 +913,36 @@ export const build = async ({
...
         const reason = reasons[file];
         if (reason && reason.type === 'initial') {
           // Initial files are manually added to the lambda later
           return;
         }
-        const { mode } = lstatSync(path.join(workPath, file));
+        const filePath = path.join(workPath, file);
+
+        if (!lstatResults[filePath]) {
+          lstatResults[filePath] = lstat(filePath);
```
Comment (Timer) [inline packages/now-next/src/index.ts]: Sema should be here, not in `Promise.all`.

Comment (ijjk) [reply packages/now-next/src/index.ts]: Added

File: packages/now-next/src/index.ts (hunk 1 above)
Comment (ijjk) [reply packages/now-next/src/index.ts:947]: Added

File: packages/now-next/src/index.ts [hunk 4]
```diff
@@ -913,25 +914,43 @@ export const build = async ({
...
-      ) => (file: string) => {
+      ) => async (file: string) => {
         const reason = reasons[file];
         if (reason && reason.type === 'initial') {
           // Initial files are manually added to the lambda later
           return;
         }
-        const { mode } = lstatSync(path.join(workPath, file));
+        const filePath = path.join(workPath, file);
+        let acquiredSema = false;
```
Comment (Timer) [inline packages/now-next/src/index.ts]:
```suggestion
```

File: packages/now-next/src/index.ts [hunk 5]
```diff
@@ -913,25 +914,43 @@ export const build = async ({
...
+
+        if (!lstatResults[filePath]) {
+          await lstatSema.acquire();
+          acquiredSema = true;
```
Comment (Timer) [inline packages/now-next/src/index.ts]:
```suggestion
```

File: packages/now-next/src/index.ts [hunk 6]
```diff
@@ -913,25 +914,43 @@ export const build = async ({
...
+          lstatResults[filePath] = lstat(filePath);
+        }
+        const { mode } = await lstatResults[filePath];
+        if (acquiredSema) lstatSema.release();
```
Comment (Timer) [inline packages/now-next/src/index.ts]:
```suggestion
```

File: packages/now-next/src/index.ts [hunk 7]
```diff
@@ -913,25 +914,43 @@ export const build = async ({
...
           return;
         }
-        const { mode } = lstatSync(path.join(workPath, file));
+        const filePath = path.join(workPath, file);
+        let acquiredSema = false;
+
+        if (!lstatResults[filePath]) {
+          await lstatSema.acquire();
+          acquiredSema = true;
+          lstatResults[filePath] = lstat(filePath);
```
Comment (Timer) [inline packages/now-next/src/index.ts]:
```suggestion
          lstatResults[filePath] = lstatSema.acquire().then(() => lstat(filePath)).finally(lstatSema.release);
```

Comment (ijjk) [reply packages/now-next/src/index.ts]: Much cleaner, updated 👍

**Review by Timer — APPROVED** [review]

**Review by timneutkens — APPROVED** [review]

**Pull request merged**